src/util/codec/mysql/decimal.rs

```diff
@@ -0,0 +1,679 @@
+        fsp = cmp::min(fsp, MAX_FSP);
+
+        let d = Decimal {
+            coeff: coeff,
+            exp: exp,
+            fsp: fsp,
+        };
+
+        if exp >= 0 {
+            return d;
+        }
+
+        if fsp as i32 + exp < 0 {
+            return d.rescale(-(fsp as i32));
+        }
+        d
+    }
+
+    /// Remove extra zero in `coeff`.
+    ///
+    /// For example, 1000 * 10<sup>3</sup> will be compact to 1 * 10<sup>6</sup>,
+    /// 1000 * 10<sup>-3</sup> will be compact to 1 * 10<sup>0</sup>.
+    pub fn compact(mut self) -> Decimal {
+        let ten = 10.into();
+        loop {
+            let res = self.coeff.div_rem(&ten);
+            if !res.1.is_zero() {
+                break;
+            }
+            self.coeff = res.0;
+            self.exp += 1;
+        }
+        self
+    }
+
+    /// Adjust decimal's exponent to specified value.
+    ///
+    /// This may cause the coefficient grows or decreases.
+    /// For example, 12345 * 10<sup>3</sup> rescale with 4 will be represented
+    /// as 1235 * 10<sup>4</sup>; 12345 * 10<sup>-3</sup> rescale with -5 will
+    /// be represented as 1234500 * 10<sup>-5</sup>
+    pub fn rescale(&self, mut exp: i32) -> Decimal {
+        if exp < 0 {
+            exp = cmp::max(exp, -(MAX_FSP as i32));
+        }
+        if self.exp == exp {
+            return self.clone();
+        }
+
+        let to_add = exp - self.exp;
+        let coeff = if to_add > 0 {
+            let divider = BigInt::from(num::pow(BigUint::from(10u8), to_add as usize));
+            let (mut coeff, rem) = self.coeff.div_rem(&divider);
+            if rem >= divider / BigInt::from(2) {
+                coeff = coeff + BigInt::one();
+            }
+            coeff
+        } else {
+            self.coeff.clone() * BigInt::from(num::pow(BigUint::from(10u8), -to_add as usize))
+        };
+
+        if coeff.is_zero() && exp > 0 {
+            exp = 0;
+        }
+
+        Decimal {
+            coeff: coeff,
+            exp: exp,
+            fsp: if exp < 0 {
+                -exp as usize
+            } else {
+                0
+            },
+        }
+    }
+
+    /// Convert the decimal to float value.
+    ///
+    /// Please note that this convertion may lose precision.
+    pub fn to_f64(&self) -> Result<f64> {
+        let s = format!("{}", self);
+        // Can this line really return error?
+        let f = box_try!(s.parse::<f64>());
+        Ok(f)
+    }
+
+    /// Test if the decimal is negative.
+    pub fn is_negative(&self) -> bool {
+        self.coeff.is_negative()
+    }
+
+    /// Test if the decimal is zero.
+    pub fn is_zero(&self) -> bool {
+        self.coeff.is_zero()
+    }
+
+    /// Get the max needed capacity to encode this decimal.
+    ///
+    /// see also `encode_decimal`.
+    pub fn max_encode_bytes(&self) -> usize {
+        let bits_cnt = self.coeff.bits();
+
+        // 2 * 2 * 2 ≈ 10
+        let number_cnt = (bits_cnt + 2) / 3;
```

Comment:
Actually it's a little trick to implement `bits_cnt / 3 + if bits_cnt % 3 > 0 {1} else {0}`.